File: app/src/main/java/com/emiryanvl/tapreader/data/repositories/BookRepositoryImpl.java
package com.emiryanvl.tapreader.data.repositories;

import com.emiryanvl.tapreader.data.dao.BookDao;
import com.emiryanvl.tapreader.data.entities.BookEntity;
import com.emiryanvl.tapreader.domain.model.Book;
import com.emiryanvl.tapreader.domain.repositories.BookRepository;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;

public class BookRepositoryImpl implements BookRepository {

    BookDao bookDao;

    @Inject
    public BookRepositoryImpl(BookDao bookDao) {
        this.bookDao = bookDao;
    }

    @Override
    public Observable<List<Book>> getAllBooks() {
        return bookDao.findAll()
            .map(bookEntities -> {
                return bookEntities.stream()
                    .map(BookEntity::toBookModel)
                    .collect(Collectors.toList());
            });
    }

    @Override
    public Observable<Book> getBook(int id) {
        return bookDao.findById(id).map(bookEntity -> bookEntity.toBookModel());
    }

    @Override
    public Completable addBook(Book book) {
        return bookDao.insert(
            new BookEntity(
                book.getTitle(),
                book.getAuthor(),
                book.getDescription(),
                book.getGenre()
            )
        );
    }

    @Override
    public void updateBook(int id, Book book) {
        BookEntity bookEntity = bookDao.findById(id).blockingFirst();
        bookEntity.setTitle(book.getTitle());
        bookEntity.setAuthor(book.getAuthor());
        bookEntity.setDescription(book.getDescription());
        bookEntity.setGenre(book.getGenre());
    }

    @Override
    public void deleteBook(int id) {
        BookEntity book = bookDao.findById(id).blockingFirst();
        bookDao.delete(book);
    }
}

File: app/src/main/java/com/emiryanvl/tapreader/data/repositories/BookRepositoryImpl.kt
package com.emiryanvl.tapreader.data.repositories

import com.emiryanvl.tapreader.data.dao.BookDao
import com.emiryanvl.tapreader.data.entities.BookEntity
import com.emiryanvl.tapreader.domain.models.Book
import com.emiryanvl.tapreader.domain.repositories.BookRepository
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class BookRepositoryImpl @Inject constructor(private val dao: BookDao) : BookRepository {

    override fun getAllBooks() = dao.findAllBooks().map { it.map(BookEntity::toBookModel) }

    override suspend fun getBook(id: Int) = BookEntity.toBookModel(dao.findBookById(id))

    override suspend fun addBook(book: Book) = dao.insertBook(BookEntity.toBookEntity(book))

    override suspend fun updateBook(id: Int, book: Book) = dao.insertBook(
        book = dao.findBookById(id).copy(
            title = book.title,
            author = book.author,
            description = book.description,
            genre = book.genre
        )
    )

    override suspend fun deleteBook(id: Int) = dao.deleteBookById(
        book = dao.findBookById(id)
    )
}